File: app/src/main/java/com/oguz/spy/domain/GameStateManager.kt
package com.oguz.spy.domain

import com.oguz.spy.ux.Category
import com.oguz.spy.ux.GamePlayer
import com.oguz.spy.ux.Player

class GameStateManager {

    private var currentGameState: GameState = GameState()

    fun updateSettings(playerCount: Int, gameDuration: Int, showHints: Boolean) {
        currentGameState = currentGameState.copy(
            playerCount = playerCount,
            gameDurationMinutes = gameDuration,
            showHints = showHints
        )
    }

    fun setPlayers(players: List<Player>) {
        currentGameState = currentGameState.copy(players = players)
    }

    fun startGame(category: Category, gamePlayers: List<GamePlayer>) {
        currentGameState = currentGameState.copy(
            selectedCategory = category,
            gamePlayers = gamePlayers,
            currentPhase = GamePhase.GAME_STARTED
        )
    }

    fun moveToPhase(phase: GamePhase) {
        currentGameState = currentGameState.copy(currentPhase = phase)
    }

    fun resetGame() {
        currentGameState = currentGameState.copy(
            selectedCategory = null,
            gamePlayers = emptyList(),
            currentPhase = GamePhase.SETUP
        )
    }

    fun getCurrentState(): GameState = currentGameState

    data class GameState(
        val playerCount: Int = 4,
        val gameDurationMinutes: Int = 5,
        val showHints: Boolean = true,
        val players: List<Player> = emptyList(),
        val selectedCategory: Category? = null,
        val gamePlayers: List<GamePlayer> = emptyList(),
        val currentPhase: GamePhase = GamePhase.SETUP
    )

    enum class GamePhase {
        SETUP,
        PLAYER_SETUP,
        CATEGORY_SELECTION,
        GAME_STARTED,
        TIMER_RUNNING,
        VOTING,
        RESULTS
    }
}

File: react-native-spy/src/domain/GameStateManager.ts
import {
  GameState,
  GamePhase,
  GameSettings,
  Player,
  GamePlayer,
  Category,
  Subcategory,
} from '../types';

export class GameStateManager {
  private state: GameState;

  constructor(initialState?: Partial<GameState>) {
    this.state = {
      currentPhase: 'setup',
      settings: {
        playerCount: 4,
        gameDurationMinutes: 5,
        showHints: true,
      },
      players: [],
      gamePlayers: [],
      selectedCategory: null,
      selectedSubcategory: null,
      timerStartTime: null,
      votes: {},
      ...initialState,
    };
  }

  getState(): GameState {
    return { ...this.state };
  }

  updateSettings(settings: Partial<GameSettings>): GameState {
    this.state = {
      ...this.state,
      settings: {
        ...this.state.settings,
        ...settings,
      },
    };
    return this.getState();
  }

  setPlayers(players: Player[]): GameState {
    this.state = {
      ...this.state,
      players,
    };
    return this.getState();
  }

  setGamePlayers(gamePlayers: GamePlayer[]): GameState {
    this.state = {
      ...this.state,
      gamePlayers,
    };
    return this.getState();
  }

  selectCategory(
    category: Category,
    subcategory?: Subcategory
  ): GameState {
    this.state = {
      ...this.state,
      selectedCategory: category,
      selectedSubcategory: subcategory || null,
    };
    return this.getState();
  }

  startGame(category: Category, gamePlayers: GamePlayer[]): GameState {
    this.state = {
      ...this.state,
      currentPhase: 'game',
      selectedCategory: category,
      gamePlayers,
      timerStartTime: null,
      votes: {},
    };
    return this.getState();
  }

  startTimer(): GameState {
    this.state = {
      ...this.state,
      currentPhase: 'timer',
      timerStartTime: Date.now(),
    };
    return this.getState();
  }

  startVoting(): GameState {
    this.state = {
      ...this.state,
      currentPhase: 'voting',
      votes: {},
    };
    return this.getState();
  }

  recordVote(voterId: string, votedPlayerId: string): GameState {
    this.state = {
      ...this.state,
      votes: {
        ...this.state.votes,
        [voterId]: votedPlayerId,
      },
    };
    return this.getState();
  }

  moveToPhase(phase: GamePhase): GameState {
    this.state = {
      ...this.state,
      currentPhase: phase,
    };
    return this.getState();
  }

  resetGame(): GameState {
    this.state = {
      ...this.state,
      currentPhase: 'setup',
      gamePlayers: [],
      selectedCategory: null,
      selectedSubcategory: null,
      timerStartTime: null,
      votes: {},
    };
    return this.getState();
  }

  canStartGame(): boolean {
    return (
      this.state.players.length >= 3 &&
      this.state.selectedCategory !== null
    );
  }

  canStartTimer(): boolean {
    return this.state.gamePlayers.length >= 3;
  }

  canStartVoting(): boolean {
    return this.state.gamePlayers.length >= 3;
  }

  getElapsedTime(): number {
    if (!this.state.timerStartTime) return 0;
    return Math.floor((Date.now() - this.state.timerStartTime) / 1000);
  }
}
